Lesson8

JuicyFlash opened the pull request:
Реализованы стандартные действия и представления для ресурса вопросов
Реализовано представление для ресурса тестов (с количеством вопросов по каждому тесту)
Добавлен футер по заданию
Добавлены хелперы по заданию

JuicyFlash committed:
стандартные действия вместе с представлениями (видами) для ресурса вопросов

JuicyFlash committed:
представление тестов

JuicyFlash committed:
футер и навигационная панель

JuicyFlash committed:
хелперы

BubuntuClu commented on app/controllers/questions_controller.rb:
```diff
@@ -19,16 +19,32 @@ def show
+  def edit; end
+
+  def update
+    question = find_question
```
почему это не через колбэк сделано?

BubuntuClu commented on app/controllers/questions_controller.rb:
```diff
@@ -19,16 +19,32 @@ def show
+    if question.update(question_params)
+      redirect_to test_questions_path
+    else
+      render :new
```
тут рендерить надо редактирование

BubuntuClu commented on app/controllers/questions_controller.rb:
```diff
@@ -19,16 +19,32 @@ def show
   # /questions/:id
   def destroy
     @question.destroy
+    redirect_to test_questions_path
```
стоит указать редирект на тест удаленного вопроса

BubuntuClu commented on app/controllers/questions_controller.rb:
```diff
@@ -2,8 +2,8 @@ class QuestionsController < ApplicationController
 
-  before_action :find_test, only: %i[index create]
-  before_action :find_question, only: %i[show destroy]
+  before_action :find_test, only: %i[index create new edit]
```
зачем тест нужен для редиктирования?

BubuntuClu commented on app/helpers/questions_helper.rb:
```diff
@@ -1,2 +1,10 @@
 module QuestionsHelper
-end
+  def question_header(action)
```
передавай сюда объект вопроса и по его состоянияю определяй какой метод использовать

BubuntuClu commented on app/views/questions/edit.html.erb:
```diff
@@ -0,0 +1,2 @@
+<h1><%= question_header(:edit) %></h1>
```
после переделки метода, эта строка тоже перейдет в форму

BubuntuClu commented on app/views/tests/index.html.erb:
```diff
@@ -0,0 +1,15 @@
+    <th>Title</th>
+    <th>Questions count</th>
+  </tr>
+  <% Test.all.each do |test| %>
```
в коде экшена получай список вопроса, не стоит выносить запросы во вьюху

BubuntuClu commented on config/routes.rb:
```diff
@@ -1,7 +1,7 @@
   # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
   resources :tests do
-    resources :questions, shallow: true
+    resources :questions#, shallow: true
```
надо вернуть опцию

BubuntuClu commented on app/views/questions/index.html.erb:
```diff
@@ -1,4 +1,4 @@
-<h1><%=@test.title %></h1>
+<h1><%= @test.title %></h1>
```
эта вьюху и метод в контроллере не нужен. отображение вопросов переехать должно в tests#show

JuicyFlash committed:
shallow: true

JuicyFlash committed:
1. Список тестов - tests#index
2. Детализация теста (тест + список вопросов) - tests#show, рендерится с помощью фрагмента questions\_index

JuicyFlash committed:
В question_header передаётся объект question

JuicyFlash committed:
Фрагмент для рендеринга вопросов теста

JuicyFlash committed:
Изменения для нового хелпера

JuicyFlash committed:
Изменения для shallow: true + комментарии ревью

JuicyFlash commented:
1. Список тестов - tests#index
2. Детализация теста (тест + список вопросов) - tests#show, рендерится с помощью фрагмента questions\_index (не уверен в что корректно называть так фрагмент)
3. Фрагмент для рендеринга списка вопросов questions\_index.html.erb
4. Routes shallow: true
5. В хелпер question_header теперь передаётся объект question

BubuntuClu commented on app/controllers/tests_controller.rb:
```diff
@@ -1,2 +1,17 @@
 class TestsController < ApplicationController
+
+  before_action :find_test, only: %i[show]
```
не стоит делать колбэки ради одного экшена. ты написал гораздо больше кода чтобы найти тест, когда можно написать одну строку прямо в экшене

BubuntuClu commented on app/views/layouts/application.html.erb:
```diff
@@ -9,8 +9,18 @@
     <%= yield %>
   </body>
+
+  <footer> <%= current_year %>
```
это стоит вынести в паршиал `_footer`

BubuntuClu commented on app/views/questions/_index.html.erb:
```diff
@@ -1,5 +1,3 @@
-<h1><%=@test.title %></h1>
-
 <table>
```
лучше не делать такой паршиал. можно использовать `render @test.questions`, но тогда должен быть паршиал `_question`

BubuntuClu commented on app/views/questions/_form.erb:
```diff
@@ -0,0 +1,10 @@
+<%= form_with url: resource_path, model: @question, local: true do |form| %>
```
`Для того чтобы метод form_with смог сформировать верный URL используйте вызов form_with model: [@parent, @child]` используй `form_with model`, чтобы роут строился сам, не надо параметром прокидывать его в форму

JuicyFlash committed:
Убрал колбэки и лишние методы

JuicyFlash committed:
Добавлен фрагмент _footer

JuicyFlash committed:
Фрагмент _question + рендеринг @test.questions в tests#show

JuicyFlash committed:
form_with model: [@parent, @child]

JuicyFlash commented:
TestsController - убрал колбэки и лишние методы
Добавлен фрагмент _footer
Добавлен фрагмен _question + рендеринг @test.questions в tests#show
Фрагмент _form - убрал проброс пути через параметр, перешёл на form_with model: [@parent, @child]
